_blog/2024-02-14-writing-async-rust-traits.mdx

```diff
@@ -0,0 +1,318 @@
+    ) -> Self::Future;
+}
+```
+
+Note that while the trait itself does not use async functions, the `call` function returns a future - which is what makes it compatible. Let’s see this in action by writing a struct called `Logger` which implements `Service`:
+
+```rust
+#[derive(Clone, Copy)]
+struct Logger<S> {
+    inner: S,
+}
+
+impl<S> Logger<S> {
+    fn new(inner: S) -> Self {
+        Self { inner }
+    }
+}
+
+impl<S, B> Service<Request<B>> for Logger<S> where S: Service<Request<B>> + Clone + Send {
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        println!("processing {} {}", req.method(), req.uri().path());
+        self.inner.call(req)
+    }
+}
+```
+
+Let’s focus on the `call` function and the `Future` type. We’ve defined an associated type, and the `call` function returns this type (using implicit return). Although the generic trait bounds can make this implementation difficult for less experienced Rustaceans, essentially the trait bounds require us to make sure that `Logger` implements `Clone` as well as `Send`. This is automatically implemented as it’s a marker trait as long as the inner type implements it.
+
+However, what if we try to add some code after the `self.inner.call(req)`? 
+
+```rust
+fn call(&mut self, req: Request) -> Self::Future {
+    println!("processing {} {}", req.method(), req.uri().path());
+    let fut = self.inner.call(req);
+    println!("end processing {} {}", req.method(), req.uri().path());
+    fut
+}
+```
+
+This code technically compiles but will lead to an incorrect result as we haven’t awaited the future yet!
+
+We can fix this by awaiting the call in an async block, boxing the future, and adding more trait bounds:
+
+```rust
+impl S, B> Service<Request<B>> for Logger<S>
+where
+    S: Service<Request<B>> + Clone + Send + 'static,
+    S::Future: Send,
+    B: Send + 'static {
+    type Future = BoxFuture<'static, Result<S::Response, S::Error>>;
+    // .. rest of your types and poll function
+
+fn call(&mut self, req: Request) -> Self::Future {
+    println!("processing {} {}", req.method(), req.uri().path());
+    
+    Box::pin(async move {
+        let fut = this.call(req);
+        println!("end processing");
+        fut
+    }
+}
+```
+
+Now it works! 
+
+While this is quite powerful and allows you to skip using the heap, you need to know the exact Future you want to return. If you can’t access it, you box the `Future` on the heap and return a pointer to a trait object. Additionally, the above implementation is still quite simple compared to other more advanced implementations. Unless you’re writing advanced, high-performance applications, however, this is about as difficult as it gets for most (simple) CRUD web services.
+
+### Using async_trait
```

Comment:
I'm wondering if we should refactor this a bit, since `async fn in traits` is supported natively in Rust now, with [some caveats](https://blog.rust-lang.org/2023/12/21/async-fn-rpit-in-traits.html). The [async-trait crate docs](https://docs.rs/async-trait/0.1.77/async_trait/) have now changed to reflect that is primarily useful for supporting traits containing async fns as dyn Traits. 

Perhaps we should refactor this section to talk about the recent async fn in trait native support, and then mention the async-trait crate and when you might still want to use it?